File: years/2022/d05.py
import os.path

testing = False
day = '05'
if testing:
    f_input = os.path.join('inputs', f'input{day}-test.txt')
else:
    f_input = os.path.join('inputs', f'input{day}.txt')

def part1(input):
    stacks = {1: ['B', 'Q', 'C'],
              2: ['R', 'Q', 'W', 'Z'],
              3: ['B', 'M', 'R', 'L', 'V'],
              4: ['C', 'Z', 'H', 'V', 'T', 'W'],
              5: ['D', 'Z', 'H', 'B', 'N', 'V', 'G'],
              6: ['H', 'N', 'P', 'C', 'J', 'F', 'V', 'Q'],
              7: ['D', 'G', 'T', 'R', 'W', 'Z', 'S'],
              8: ['C', 'G', 'M', 'N', 'B', 'W', 'Z', 'P'],
              9: ['N', 'J', 'B', 'M', 'W', 'Q', 'F', 'P']}
    with open(input) as f:
        for line in f:
            if 'move' not in line:
                continue

            h1, h2 = line.strip().split(' from ')
            num = int(h1.split()[-1])
            temp = h2.split(' to ')
            start_stack = int(temp[0])
            end_stack = int(temp[1])

            for _ in range(num):
                tmp = stacks[start_stack].pop()
                stacks[end_stack].append(tmp)

    return ''.join(stacks[i][-1] for i in range(1, 10))

print(f'Part 1: {part1(f_input)}')


def part2(input):
    stacks = {1: ['B', 'Q', 'C'],
              2: ['R', 'Q', 'W', 'Z'],
              3: ['B', 'M', 'R', 'L', 'V'],
              4: ['C', 'Z', 'H', 'V', 'T', 'W'],
              5: ['D', 'Z', 'H', 'B', 'N', 'V', 'G'],
              6: ['H', 'N', 'P', 'C', 'J', 'F', 'V', 'Q'],
              7: ['D', 'G', 'T', 'R', 'W', 'Z', 'S'],
              8: ['C', 'G', 'M', 'N', 'B', 'W', 'Z', 'P'],
              9: ['N', 'J', 'B', 'M', 'W', 'Q', 'F', 'P']}
    with open(input) as f:
        for line in f:
            if 'move' not in line:
                continue
            h1, h2 = line.strip().split(' from ')
            num = int(h1.split()[-1])
            temp = h2.split(' to ')
            start_stack = int(temp[0])
            end_stack = int(temp[1])

            placeholder = []
            for _ in range(num):
                tmp = stacks[start_stack].pop()
                placeholder.append(tmp)

            placeholder.reverse()
            stacks[end_stack].extend(placeholder)

    return ''.join(stacks[i][-1] for i in range(1, 10))

print(f'Part 2: {part2(f_input)}')

File: years/2022/d05.rb
# frozen_string_literal: true

testing = false
day = '05'

if testing
  f_input = File.join('inputs', "input#{day}-test.txt")
else
  f_input = File.join('inputs', "input#{day}.txt")
end

def part1(input)
  # split input into cargo map and directions
  raw_cargo = []
  raw_directions = []
  File.foreach(input) do |line|
    line.chomp!
    if line.include? 'move'
      raw_directions << line
    elsif line.empty?
      next
    else
      raw_cargo << line
    end
  end

  # create hash that'll store the stacks of boxes
  stacks = Hash.new
  num_stacks = raw_cargo.pop.split.map(&:to_i)[-1]
  (1..num_stacks).each do |val|
    stacks[val] = []
  end

  # parse boxes into their stacks
  raw_cargo.reverse_each do |line|
    (1..num_stacks).each do |stack_num|
      index = 4 * stack_num - 3
      box = line[index]
      unless box.strip.empty?
        stacks[stack_num] << box
      end
    end
  end

  # start moving boxes according to directions
  raw_directions.each do |line|
    _move, num_boxes, _from, start_stack, _to, end_stack = line.split

    num_boxes.to_i.times do |i|
      stacks[end_stack.to_i] << stacks[start_stack.to_i].pop
    end
  end

  letters = []
  (1..num_stacks).each do |i|
    letters << stacks[i][-1]
  end

  letters.join
end

puts("Part 1: #{part1(f_input)}")


def part2(input)
  # split input into cargo map and directions
  raw_cargo = []
  raw_directions = []
  File.foreach(input) do |line|
    line.chomp!
    if line.include? 'move'
      raw_directions << line
    elsif line.empty?
      next
    else
      raw_cargo << line
    end
  end

  # create hash that'll store the stacks of boxes
  stacks = Hash.new
  num_stacks = raw_cargo.pop.split.map(&:to_i)[-1]
  (1..num_stacks).each do |val|
    stacks[val] = []
  end

  # parse boxes into their stacks
  raw_cargo.reverse_each do |line|
    (1..num_stacks).each do |stack_num|
      index = 4 * stack_num - 3
      box = line[index]
      unless box.strip.empty?
        stacks[stack_num] << box
      end
    end
  end

  # start moving boxes according to directions
  raw_directions.each do |line|
    _move, num_boxes, _from, start_stack, _to, end_stack = line.split

    stacks[end_stack.to_i].concat(stacks[start_stack.to_i].pop(num_boxes.to_i))
  end

  letters = []
  (1..num_stacks).each do |i|
    letters << stacks[i][-1]
  end

  letters.join
end

puts("Part 2: #{part2(f_input)}")
